engines/ultima/ultima8/graphics/shape.cpp

```diff
@@ -36,63 +37,56 @@ DEFINE_RUNTIME_CLASSTYPE_CODE_BASE_CLASS(Shape)
 
 DEFINE_CUSTOM_MEMORY_ALLOCATION(Shape)
 
-Shape::Shape(const uint8 *data_, uint32 size_, const ConvertShapeFormat *format,
-             const uint16 id, const uint32 shape) : _flexId(id), _shapeNum(shape) {
+Shape::Shape(const uint8 *data, uint32 size, const ConvertShapeFormat *format,
+             const uint16 id, const uint32 shape)
+		: _flexId(id), _shapeNum(shape), _palette(nullptr) {
 	// NB: U8 style!
+	loadFrames(data, size, format);
 
-	this->_data = data_;
-	this->_size = size_;
-	this->_palette = nullptr;
-
-	if (!format) format = DetectShapeFormat(data_, size_);
-
-	if (!format) {
-		// Should be fatal?
-		perr << "Error: Unable to detect shape format" << Std::endl;
-		return;
-	}
-
-	// Load it as u8
-	if (format == &U8ShapeFormat || format == &U82DShapeFormat)
-		LoadU8Format(data_, size_, format);
-	else if (format == &PentagramShapeFormat)
-		LoadPentagramFormat(data_, size_, format);
-	else
-		LoadGenericFormat(data_, size_, format);
+	delete[] const_cast<uint8 *>(data);
 }
 
 Shape::Shape(IDataSource *src, const ConvertShapeFormat *format)
 	: _flexId(0), _shapeNum(0), _palette(nullptr) {
 	// NB: U8 style!
+	uint32 size = src->getSize();
+	uint8 *data = new uint8[size];
+	src->read(data, size);
 
-	this->_size = src->getSize();
-	uint8 *d = new uint8[this->_size];
-	this->_data = d;
-	src->read(d, this->_size);
+	loadFrames(data, size, format);
 
+	delete[] data;
+}
+
+Shape::~Shape() {
+	for (unsigned int i = 0; i < _frames.size(); ++i)
+		delete _frames[i];
+}
+
+void Shape::loadFrames(const uint8 *data, uint32 size, const ConvertShapeFormat *format) {
 	if (!format)
-		format = DetectShapeFormat(_data, _size);
+		format = DetectShapeFormat(data, size);
 
 	if (!format) {
 		// Should be fatal?
 		perr << "Error: Unable to detect shape format" << Std::endl;
 		return;
 	}
 
+	Std::vector<RawShapeFrame *> rawframes;
```

Comment:
Since it is a new code, consider using `Common::Array` directly and avoid the wrapper.

Reply:
Done, another thing that I should go through and clean up everywhere.